File: web/app/Application/Proveedores/VincularUsuarioProveedorUseCase.php
<?php

namespace App\Application\Proveedores;

use App\Domain\Proveedores\Exceptions\ProveedorInvalidoException;
use App\Domain\Proveedores\Proveedor;
use App\Domain\Proveedores\ProveedorRepositoryInterface;
use App\Domain\Usuarios\Rol;
use App\Domain\Usuarios\UsuarioRepositoryInterface;
use RuntimeException;

final class VincularUsuarioProveedorUseCase
{
    public function __construct(
        private readonly ProveedorRepositoryInterface $proveedores,
        private readonly UsuarioRepositoryInterface $usuarios,
    ) {}

    public function ejecutar(int $proveedorId, int $usuarioId): Proveedor
    {
        $proveedor = $this->proveedores->buscarPorId($proveedorId);
        if ($proveedor === null) {
            throw new RuntimeException('El proveedor no existe.');
        }

        $usuario = $this->usuarios->buscarPorId($usuarioId);
        if ($usuario === null || ! $usuario->tieneRol(Rol::Proveedor)) {
            throw ProveedorInvalidoException::usuarioSinRolProveedor();
        }

        $vinculadoActual = $this->proveedores->buscarPorUsuarioId($usuarioId);
        if ($vinculadoActual !== null && $vinculadoActual->id !== $proveedor->id) {
            throw ProveedorInvalidoException::usuarioYaVinculado();
        }

        return $this->proveedores->guardar($proveedor->vincularUsuario($usuarioId));
    }
}

File: shared/src/commonMain/kotlin/pe/ecolecta/domain/usecase/proveedor/VincularUsuarioProveedorUseCase.kt
package pe.ecolecta.domain.usecase.proveedor

import pe.ecolecta.domain.ProveedorInvalidoException
import pe.ecolecta.domain.model.Rol
import pe.ecolecta.domain.repository.ProveedorRepository
import pe.ecolecta.domain.repository.UsuarioRepository

/**
 * Vincula una cuenta de Usuario (rol PROVEEDOR) a un registro Proveedor, para que
 * [pe.ecolecta.domain.usecase.proveedor.ObtenerProveedorAsociadoUseCase] pueda resolverlo al iniciar
 * sesión. Sin este vínculo, un proveedor creado por el ADMIN nunca puede ver su perfil ni sus entregas.
 */
class VincularUsuarioProveedorUseCase(
    private val proveedorRepository: ProveedorRepository,
    private val usuarioRepository: UsuarioRepository,
) {
    suspend operator fun invoke(proveedorId: String, usuarioId: String): Result<Unit> = runCatching {
        val usuario = usuarioRepository.obtenerPorId(usuarioId)
            ?: throw NoSuchElementException("El usuario no existe.")
        if (Rol.PROVEEDOR !in usuario.roles) {
            throw ProveedorInvalidoException.UsuarioSinRolProveedor
        }
        proveedorRepository.obtenerPorId(proveedorId)
            ?: throw NoSuchElementException("El proveedor no existe.")

        val yaVinculadoA = proveedorRepository.obtenerPorUsuarioId(usuarioId)
        if (yaVinculadoA != null && yaVinculadoA.id != proveedorId) {
            throw ProveedorInvalidoException.UsuarioYaVinculado
        }

        proveedorRepository.vincularUsuario(proveedorId, usuarioId)
    }
}
